File: packages/python/elizaos/features/basic_capabilities/providers/action_state.py
from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State


async def get_action_state_context(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    sections: list[str] = []
    action_data: dict[str, list[str]] = {
        "pending": [],
        "completed": [],
        "available": [],
    }

    # Use the actions property instead of get_available_actions()
    available_actions = runtime.actions
    action_data["available"] = [a.name for a in available_actions]

    if action_data["available"]:
        sections.append("## Available Actions")
        sections.append(", ".join(action_data["available"]))

    if state and state.values:
        # Handle both dict-like and protobuf state.values
        values = state.values
        pending = []
        completed = []

        if hasattr(values, "get") and callable(values.get):
            # Dict-like access
            pending = values.get("pendingActions", [])
            completed = values.get("completedActions", [])
        elif hasattr(values, "extra"):
            # Protobuf - extra might be a MapField or a message
            extra = values.extra
            if hasattr(extra, "get") and callable(extra.get):
                pending_raw = extra.get("pendingActions", "")
                completed_raw = extra.get("completedActions", "")
            elif hasattr(extra, "__getitem__"):
                try:
                    pending_raw = extra["pendingActions"]
                except (KeyError, TypeError, ValueError):
                    pending_raw = ""
                try:
                    completed_raw = extra["completedActions"]
                except (KeyError, TypeError, ValueError):
                    completed_raw = ""
            else:
                pending_raw = ""
                completed_raw = ""

            # Parse if stored as string
            if isinstance(pending_raw, str) and pending_raw:
                import json

                try:
                    pending = json.loads(pending_raw)
                except json.JSONDecodeError:
                    pending = []
            elif isinstance(pending_raw, list):
                pending = pending_raw

            if isinstance(completed_raw, str) and completed_raw:
                import json

                try:
                    completed = json.loads(completed_raw)
                except json.JSONDecodeError:
                    completed = []
            elif isinstance(completed_raw, list):
                completed = completed_raw

        if isinstance(pending, list):
            action_data["pending"] = [str(a) for a in pending]

        if isinstance(completed, list):
            action_data["completed"] = [str(a) for a in completed]

    if action_data["pending"]:
        sections.append("\n## Pending Actions")
        sections.append("\n".join(f"- {a}" for a in action_data["pending"]))

    if action_data["completed"]:
        sections.append("\n## Recently Completed")
        sections.append("\n".join(f"- {a}" for a in action_data["completed"][-5:]))

    context_text = "# Action State\n" + "\n".join(sections) if sections else ""

    return ProviderResult(
        text=context_text,
        values={
            "availableActionCount": len(action_data["available"]),
            "pendingActionCount": len(action_data["pending"]),
            "completedActionCount": len(action_data["completed"]),
        },
        data=action_data,
    )


action_state_provider = Provider(
    name="ACTION_STATE",
    description="Provides information about the current action state and available actions",
    get=get_action_state_context,
    dynamic=True,
)

File: packages/rust/src/features/basic_capabilities/providers/action_state.rs
//! ACTION_STATE provider implementation.

use async_trait::async_trait;
use once_cell::sync::Lazy;

use crate::error::PluginResult;
use crate::generated::spec_helpers::require_provider_spec;
use crate::runtime::IAgentRuntime;
use crate::types::{Memory, ProviderResult, State};

use super::Provider;

static SPEC: Lazy<&'static crate::generated::spec_helpers::ProviderDoc> =
    Lazy::new(|| require_provider_spec("ACTION_STATE"));

/// Provider for action state information.
pub struct ActionStateProvider;

#[async_trait]
impl Provider for ActionStateProvider {
    fn name(&self) -> &'static str {
        &SPEC.name
    }

    fn description(&self) -> &'static str {
        &SPEC.description
    }

    fn is_dynamic(&self) -> bool {
        SPEC.dynamic.unwrap_or(true)
    }

    async fn get(
        &self,
        runtime: &dyn IAgentRuntime,
        _message: &Memory,
        state: Option<&State>,
    ) -> PluginResult<ProviderResult> {
        let mut sections = Vec::new();

        // Get available actions
        let available_actions = runtime.get_available_actions();
        let action_names: Vec<String> = available_actions.iter().map(|a| a.name.clone()).collect();

        if !action_names.is_empty() {
            sections.push("## Available Actions".to_string());
            sections.push(action_names.join(", "));
        }

        // Get action state from state if available
        let mut pending: Vec<String> = Vec::new();
        let mut completed: Vec<String> = Vec::new();

        if let Some(state) = state {
            if let Some(pending_val) = state.get_value("pendingActions") {
                if let Some(arr) = pending_val.as_array() {
                    pending = arr
                        .iter()
                        .filter_map(|v| v.as_str().map(String::from))
                        .collect();
                }
            }

            if let Some(completed_val) = state.get_value("completedActions") {
                if let Some(arr) = completed_val.as_array() {
                    completed = arr
                        .iter()
                        .filter_map(|v| v.as_str().map(String::from))
                        .collect();
                }
            }
        }

        if !pending.is_empty() {
            sections.push("\n## Pending Actions".to_string());
            sections.push(
                pending
                    .iter()
                    .map(|a| format!("- {}", a))
                    .collect::<Vec<_>>()
                    .join("\n"),
            );
        }

        if !completed.is_empty() {
            sections.push("\n## Recently Completed".to_string());
            let recent: Vec<_> = completed.iter().rev().take(5).collect();
            sections.push(
                recent
                    .iter()
                    .map(|a| format!("- {}", a))
                    .collect::<Vec<_>>()
                    .join("\n"),
            );
        }

        let context_text = if sections.is_empty() {
            String::new()
        } else {
            format!("# Action State\n{}", sections.join("\n"))
        };

        Ok(ProviderResult::new(context_text)
            .with_value("availableActionCount", action_names.len() as i64)
            .with_value("pendingActionCount", pending.len() as i64)
            .with_value("completedActionCount", completed.len() as i64)
            .with_data("available", action_names)
            .with_data("pending", pending)
            .with_data("completed", completed))
    }
}
